crates/ruma-events/src/file.rs

```diff
@@ -0,0 +1,224 @@
+    pub fn is_encrypted(&self) -> bool {
+        self.encryption_info.is_some()
+    }
+}
+
+/// Trait for messages containing a file.
+pub trait FileMessage {
+    /// Get the file of the message.
+    fn file(&self) -> &FileContent;
+}
+
+/// The payload for an extensible text message.
+#[derive(Clone, Debug, Serialize, Deserialize, EventContent)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[ruma_event(type = "m.file", kind = MessageLike)]
+pub struct FileEventContent {
+    /// The text representation of the message.
+    #[serde(flatten)]
+    pub message: MessageContent,
+
+    /// The file content of the message.
+    #[serde(rename = "org.matrix.msc1767.file")]
+    pub file: FileContent,
+
+    /// Information about related messages for [rich replies].
+    ///
+    /// [rich replies]: https://spec.matrix.org/v1.2/client-server-api/#rich-replies
+    #[serde(flatten, skip_serializing_if = "Option::is_none")]
+    pub relates_to: Option<Relation>,
+}
+
+impl FileEventContent {
+    /// Creates a new non-encrypted `FileEventContent` with the given plain text message, url and
+    /// file info.
+    pub fn plain(
+        message: impl Into<String>,
+        url: Box<MxcUri>,
+        info: Option<Box<FileContentInfo>>,
+    ) -> Self {
+        Self {
+            message: MessageContent::plain(message),
+            file: FileContent::plain(url, info),
+            relates_to: None,
+        }
+    }
+
+    /// Creates a new non-encrypted `FileEventContent` with the given message, url and
+    /// file info.
+    pub fn plain_message(
+        message: MessageContent,
+        url: Box<MxcUri>,
+        info: Option<Box<FileContentInfo>>,
+    ) -> Self {
+        Self { message, file: FileContent::plain(url, info), relates_to: None }
+    }
+
+    /// Creates a new encrypted `FileEventContent` with the given plain text message, url,
+    /// encryption info and file info.
+    pub fn encrypted(
+        message: impl Into<String>,
+        url: Box<MxcUri>,
+        encryption_info: EncryptedContent,
+        info: Option<Box<FileContentInfo>>,
+    ) -> Self {
+        Self {
+            message: MessageContent::plain(message),
+            file: FileContent::encrypted(url, encryption_info, info),
+            relates_to: None,
+        }
+    }
+
+    /// Creates a new encrypted `FileEventContent` with the given message, url,
+    /// encryption info and file info.
+    pub fn encrypted_message(
+        message: MessageContent,
+        url: Box<MxcUri>,
+        encryption_info: EncryptedContent,
+        info: Option<Box<FileContentInfo>>,
+    ) -> Self {
+        Self { message, file: FileContent::encrypted(url, encryption_info, info), relates_to: None }
+    }
+}
+
+impl TextMessage for FileEventContent {
+    fn message(&self) -> &MessageContent {
+        &self.message
+    }
+}
+
+impl FileMessage for FileEventContent {
+    fn file(&self) -> &FileContent {
+        &self.file
+    }
+}
```

Comment:
Oh, I didn't review carefully enough last time! I'm still not really convinced these traits are that useful.

Let me try to explain: I don't think anybody would want to write a function that's generic over (just) known event types that definitely contain a file / message / whatever. Instead, you would want to have a map of known extensible event content types to some logic for handling them, regardless of whether the overall event type are known in the version of Ruma being used.

Reply:
It's true that I was wondering how to handle custom events with extensible types. I'm still not sure how that would go but for now I'll trust your judgement and remove these traits.